### PR TITLE
Bump Merkle Proof to use depth of 8

### LINKED ISSUE #26: Increase Binary Merkle Proof To Depth of 8
#### Description

The Binary Merkle Proof Library currently tests with a depth of 3. This should be increased to a depth of 8.

#### Motiviation

The current tests use an array of length 3. The standard for testing Merkle trees is a depth of 8.

### PR DESCRIPTION
**Pull request by bitzoic**
## Type of change

- Improvement (refactoring, restructuring repository, cleaning tech debt, ...)

## Changes

The following changes have been made:

- Use a fixed array of length 8 for binary Merkle proof
- Function to automatically return leaves for a perfect binary tree
- Function to generate a Merkle proof for a perfect binary tree

## Related Issues

Closes #26

### CODE REVIEW

**Comment by simonr0204** [comment]
I may be out of the loop here, but obvious question: why 8?  That's still only 256 leaves. Are smaller trees supported (with empty values in the proof array) ?

**Comment by bitzoic** [comment]
> I may be out of the loop here, but obvious question: why 8? That's still only 256 leaves. Are smaller trees supported (with empty values in the proof array) ?

It was requested by @SilentCicero to bump the depth to the standard of 8. 

It was previously only a depth of 3.

**Comment by simonr0204** [comment]
> > I may be out of the loop here, but obvious question: why 8? That's still only 256 leaves. Are smaller trees supported (with empty values in the proof array) ?
> 
> It was requested by @SilentCicero to bump the depth to the standard of 8.
> 
> It was previously only a depth of 3.

Just to check - fundamentally this is just because `Vec` is not supported yet, right? The term "standard depth" has me a little confused.

**Comment by bitzoic** [comment]
> > > I may be out of the loop here, but obvious question: why 8? That's still only 256 leaves. Are smaller trees supported (with empty values in the proof array) ?
> > 
> > 
> > It was requested by @SilentCicero to bump the depth to the standard of 8.
> > It was previously only a depth of 3.
> 
> Just to check - fundamentally this is just because `Vec` is not supported yet, right? The term "standard depth" has me a little confused.

Yes `Vec` is not yet supported by the SDK so a fixed length array is needed to be used. I would like to add test cases for multiple proof sizes when they are supported. It's not so much a standard but a better test as far as I'm aware.

I am also building the tree manually as the `sha256` in Sway pads `u8`as to be the same size as `u64`. I created an issue for it here: https://github.com/FuelLabs/sway/issues/2594

File: sway_libs/src/merkle_proof/binary_merkle_proof.sw [hunk 1]
```diff
@@ -98,10 +98,10 @@ pub fn process_proof(
     key: u64,
     merkle_leaf: b256,
     num_leaves: u64,
-    proof: [b256; 2],
+    proof: [b256; 8],
```
Comment (matt-user) [inline sway_libs/src/merkle_proof/binary_merkle_proof.sw:101]: Why was this changed from 2->8 instead of 3->8?

**Review by matt-user — APPROVED** [review]
lgtm, just have a question regarding the change in proof depth

**Review by Braqzen — APPROVED** [review]
Concept ACK

File: sway_libs/src/merkle_proof/binary_merkle_proof.sw (hunk 1 above)
Comment (bitzoic) [reply sway_libs/src/merkle_proof/binary_merkle_proof.sw:101]: The tree I previously had in place wasn't a perfect binary Tree and only required 2 proofs.
